src/sql/workbench/parts/notebook/outputs/notebookMarkdown.ts

```diff
@@ -0,0 +1,240 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+import * as path from 'path';
+
+import { URI } from 'vs/base/common/uri';
+
+import { IDisposable, dispose } from 'vs/base/common/lifecycle';
+import { RenderOptions } from 'vs/base/browser/htmlContentRenderer';
+import { IMarkdownString, removeMarkdownEscapes } from 'vs/base/common/htmlContent';
+import { IMarkdownRenderResult } from 'vs/editor/contrib/markdown/markdownRenderer';
+import marked = require('vs/base/common/marked/marked');
+import { defaultGenerator } from 'vs/base/common/idGenerator';
+import { revive } from 'vs/base/common/marshalling';
+import * as fs from 'fs';
+
+export class NotebookMarkdownRenderer {
+	private _notebookURI: URI;
+	private _baseUrls: string[] = [];
+
+	constructor() {
+	}
+
+	render(markdown: IMarkdownString): IMarkdownRenderResult {
+		let disposables: IDisposable[] = [];
+		const element: HTMLElement = markdown ? this.renderMarkdown(markdown, undefined) : document.createElement('span');
+		return {
+			element,
+			dispose: () => dispose(disposables)
+		};
+	}
+
+	createElement(options: RenderOptions): HTMLElement {
+		const tagName = options.inline ? 'span' : 'div';
+		const element = document.createElement(tagName);
+		if (options.className) {
+			element.className = options.className;
+		}
+		return element;
+	}
+
+	parse(text: string): any {
+		let data = JSON.parse(text);
+		data = revive(data, 0);
+		return data;
+	}
+
+	/**
+	 * Create html nodes for the given content element.
+	 * Adapted from htmlContentRenderer. Ensures that the markdown renderer
+	 * gets passed in the correct baseUrl for the notebook's saved location,
+	 * respects the trusted state of a notebook, and allows command links to
+	 * be clickable.
+	 */
+	renderMarkdown(markdown: IMarkdownString, options: RenderOptions = {}): HTMLElement {
+		const element = this.createElement(options);
+
+		// signal to code-block render that the element has been created
+		let signalInnerHTML: () => void;
+		const withInnerHTML = new Promise(c => signalInnerHTML = c);
+
+		let notebookFolder = path.join(path.dirname(this._notebookURI.fsPath), path.sep);
+		if (!this._baseUrls.includes(notebookFolder)) {
+			this._baseUrls.push(notebookFolder);
+		}
+		const renderer = new marked.Renderer({ baseUrl: notebookFolder });
+		renderer.image = (href: string, title: string, text: string) => {
+			href = this.cleanUrl(!markdown.isTrusted, notebookFolder, href);
```

Comment:
Minor: all these are large enough they warrant their own method/function for simplicity. I know VSCode doesn't do that, but can we?